[quant][graph] Add support for FP16 dynamic quant

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42348 [quant] Use PlaceholderObserver instead of Fp16Observer and NoopObserver
* **#42222 [quant][graph] Add support for FP16 dynamic quant**
* #42221 [quant] Add FP16Observer for fp16 quant support
* #42147 [quant] Add saturate_to_fp16 op for FP16 quant support

Summary:
This change adds the necessary passes to perform FP16 dynamic quantization.
We skip inserting observers for activations based on the dtype (torch.float16) and only insert the Fp16Observer for weights

Test Plan:
python test/test_quantization.py TestQuantizeJitOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22849220](https://our.internmc.facebook.com/intern/diff/D22849220)

supriyar committed:
[quant][graph] Add support for FP16 dynamic quant
Summary:
This change adds the necessary passes to perform FP16 dynamic quantization.
We skip inserting observers for activations based on the dtype (torch.float16) and only insert the Fp16Observer for weights

Test Plan:
python test/test_quantization.py TestQuantizeJitOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add support for FP16 dynamic quant"


Summary:
This change adds the necessary passes to perform FP16 dynamic quantization.
We skip inserting observers for activations based on the dtype (torch.float16) and only insert the Fp16Observer for weights

Test Plan:
python test/test_quantization.py TestQuantizeJitOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add support for FP16 dynamic quant"


Summary:
This change adds the necessary passes to perform FP16 dynamic quantization.
We skip inserting observers for activations based on the dtype (torch.float16) and only insert the Fp16Observer for weights

Test Plan:
python test/test_quantization.py TestQuantizeJitOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -405,7 +394,7 @@ void insertQuantizationOps(
   }
   if (quant_type == QuantType::DYNAMIC) {
-    if (isFP16NoopObserver(module, observer)) {
+    if (isFp16Observer(observer->input(0))) {
```
can we just check dtype here?

supriyar replied:
I think so, was just being more explicit by checking for the observer type as well.

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -245,12 +232,14 @@ bool isNoopObserver(Value* observer) {
-  auto observer_module = module.attr(findObserverName(v).value()).toModule();
-  return (observer_module.attr("dtype") == at::ScalarType::Half) &&
-      isNoopObserver(observer);
+bool isFp16Observer(Value* observer) {
```
do we need this check? I think checking dtype is enough for our purposes?

supriyar committed:
Update on "[quant][graph] Add support for FP16 dynamic quant"


Summary:
This change adds the necessary passes to perform FP16 dynamic quantization.
We skip inserting observers for activations based on the dtype (torch.float16) and only insert the Fp16Observer for weights

Test Plan:
python test/test_quantization.py TestQuantizeJitOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -405,7 +394,9 @@ void insertQuantizationOps(
   }
   if (quant_type == QuantType::DYNAMIC) {
-    if (isFP16NoopObserver(module, observer)) {
+    if (getObserverDtype(module, observer_out).has_value() &&
```
nit: probably better to assign the result to a variable

jerryzh168 commented on torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:
```diff
@@ -245,12 +232,14 @@ bool isNoopObserver(Value* observer) {
-  auto observer_module = module.attr(findObserverName(v).value()).toModule();
-  return (observer_module.attr("dtype") == at::ScalarType::Half) &&
-      isNoopObserver(observer);
+c10::optional<at::ScalarType> getObserverDtype(Module& module, Value* v) {
```
also maybe we can return at::ScalarType?

jerryzh168 commented on test/quantization/test_quantize_jit.py:
```diff
@@ -2506,20 +2506,22 @@ def forward(self, x):
+
             if qconfig == float16_dynamic_qconfig:
-                observer_name = 'NoopObserver = prim::GetAttr[name="_observer_'
+                observer_name = 'Fp16Observer = prim::GetAttr[name="_observer_'
+                FileCheck().check(observer_name) \
+                           .run(m.fc.graph)
```
Looks like this check is not very useful, what do we want to check here?

supriyar replied:
Just additional check for observer name match

jerryzh168 reviewed (approved):
LG, had a few inline comments

supriyar committed:
Update on "[quant][graph] Add support for FP16 dynamic quant"


Summary:
This change adds the necessary passes to perform FP16 dynamic quantization.
We skip inserting observers for activations based on the dtype (torch.float16) and only insert the Fp16Observer for weights

Test Plan:
python test/test_quantization.py TestQuantizeJitOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add support for FP16 dynamic quant"


Summary:
This change adds the necessary passes to perform FP16 dynamic quantization.
We skip inserting observers for activations based on the dtype (torch.float16) and only insert the Fp16Observer for weights

Test Plan:
python test/test_quantization.py TestQuantizeJitOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22849220](https://our.internmc.facebook.com/intern/diff/D22849220)

[ghstack-poisoned]
